contracts/specs/dai.spec

```diff
@@ -0,0 +1,364 @@
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that allowance behaves correctly on decreaseAllowance
+rule decreaseAllowance(address spender, uint256 value) {
+    env e;
+
+    uint256 spenderAllowance = allowance(e.msg.sender, spender);
+
+    decreaseAllowance(e, spender, value);
+
+    assert(allowance(e.msg.sender, spender) == spenderAllowance - value, "decreaseAllowance did not decrease the allowance as expected");
+}
+
+// Verify revert rules on decreaseAllowance
+rule decreaseAllowance_revert(address spender, uint256 value) {
+    env e;
+
+    uint256 spenderAllowance = allowance(e.msg.sender, spender);
+
+    decreaseAllowance@withrevert(e, spender, value);
+
+    bool revert1 = spenderAllowance - value < 0;
+    bool revert2 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Underflow did not revert");
+    assert(revert2 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that supply and balance behave correctly on mint
+rule mint(address to, uint256 value) {
+    env e;
+
+    // Save the totalSupply and sender balance before minting
+    uint256 supply = totalSupply();
+    uint256 toBalance = balanceOf(to);
+    uint256 ward = wards(e.msg.sender);
+
+    require(supply >= toBalance);
+
+    mint(e, to, value);
+
+    assert(balanceOf(to) == toBalance + value, "Mint did not increase the balance as expected");
+    assert(totalSupply() == supply + value, "Mint did not increase the supply as expected");
+}
+
+// Verify revert rules on mint
+rule mint_revert(address to, uint256 value) {
+    env e;
+
+    // Save the totalSupply and sender balance before minting
+    uint256 supply = totalSupply();
+    uint256 ward = wards(e.msg.sender);
+
+    mint@withrevert(e, to, value);
+
+    bool revert1 = ward != 1;
+    bool revert2 = supply + value > max_uint;
+    bool revert3 = to == 0 || to == currentContract;
+    bool revert4 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Lack of auth did not revert");
+    assert(revert2 => lastReverted, "Supply overflow did not revert");
+    assert(revert3 => lastReverted, "Incorrect address did not revert");
+    assert(revert4 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2 || revert3 || revert4, "Revert rules are not covering all the cases");
+}
+
+// Verify that supply and balance behave correctly on burn
+rule burn(address from, uint256 value) {
+    env e;
+
+    uint256 supply = totalSupply();
+    uint256 fromBalance = balanceOf(from);
+    uint256 allowed = allowance(from, e.msg.sender);
+    uint256 ward = wards(e.msg.sender);
+    bool senderSameAsFrom = e.msg.sender == from;
+    bool wardsEqOne = wards(e.msg.sender) == 1;
+    bool allowedEqMaxUint = allowed == max_uint;
+
+    require(supply >= fromBalance);
```

Comment:
Same question--is this strictly necessary?

Reply:
Same as above.